// idyll-components/dist/cjs/header.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Header = function (_React$PureComponent) {
  _inherits(Header, _React$PureComponent);

  function Header() {
    _classCallCheck(this, Header);

    return _possibleConstructorReturn(this, _React$PureComponent.apply(this, arguments));
  }

  Header.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        background = _props.background,
        color = _props.color;

    return _react2.default.createElement(
      'div',
      { className: 'article-header', style: { background: background, color: color } },
      _react2.default.createElement(
        'h1',
        { className: 'hed' },
        this.props.title
      ),
      this.props.subtitle && _react2.default.createElement(
        'h2',
        { className: 'dek' },
        this.props.subtitle
      ),
      this.props.author && _react2.default.createElement(
        'div',
        { className: 'byline' },
        'By: ',
        _react2.default.createElement(
          'a',
          { href: this.props.authorLink },
          this.props.author
        )
      ),
      this.props.authors ? _react2.default.createElement(
        'div',
        { className: 'byline' },
        'By:',
        ' ',
        this.props.authors.map(function (author, i) {
          if (typeof author === 'string') {
            return author;
          }
          return author.link ? _react2.default.createElement(
            'span',
            { key: author.name },
            _react2.default.createElement(
              'a',
              { href: author.link },
              author.name
            ),
            i < _this2.props.authors.length - 1 ? i === _this2.props.authors.length - 2 ? ' and ' : ', ' : ''
          ) : author.name;
        })
      ) : null,
      this.props.date && _react2.default.createElement(
        'div',
        { className: 'idyll-pub-date' },
        this.props.date
      )
    );
  };

  return Header;
}(_react2.default.PureComponent);

Header._idyll = {
  name: 'Header',
  tagType: 'closed',
  props: [{
    name: 'title',
    type: 'string',
    example: '"Article Title"'
  }, {
    name: 'subtitle',
    type: 'string',
    example: '"Article subtitle."'
  }, {
    name: 'author',
    type: 'string',
    example: '"Author Name"'
  }, {
    name: 'authorLink',
    type: 'string',
    example: '"author.website"'
  }, {
    name: 'authors',
    type: 'array',
    example: "`[{name: 'Jane Doe', link: 'https://website.com'}]`",
    description: 'An array of authors. Each element of the array can be an object or a string.'
  }, {
    name: 'date',
    type: 'string',
    example: '"December 25, 2018"',
    description: 'The publication date.'
  }, {
    name: 'background',
    type: 'string',
    example: '"blue"',
    defaultValue: '"#222"',
    description: 'The background of the header. Can pass a color or a url().'
  }, {
    name: 'color',
    type: 'string',
    example: '"#000"',
    defaultValue: '"#fff"',
    description: 'The text color of the header.'
  }]
};

exports.default = Header;
